Description: backport some support for libktorrent from trunk
 This patch is mostly a backport of upstream commit r1117656 excluding
 CMakeLists.txt changes. See the next patch for adapted version of those.
Author: Lukas Appelhans <l.appelhans@gmx.de>
Origin: backport, http://websvn.kde.org/?view=revision&revision=1117656
Forwarded: not-needed
Last-Update: 2010-05-30

--- a/kget/transfer-plugins/bittorrent/btsettingswidget.cpp
+++ b/kget/transfer-plugins/bittorrent/btsettingswidget.cpp
@@ -28,6 +28,7 @@ BTSettingsWidget::BTSettingsWidget(QWidg
     connect(torrentEdit, SIGNAL(textChanged(const QString &)), SLOT(changed()));
     connect(tempEdit, SIGNAL(textChanged(const QString &)), SLOT(changed()));
     connect(preallocBox, SIGNAL(stateChanged(int)), SLOT(changed()));
+    connect(utpBox, SIGNAL(stateChanged(int)), SLOT(changed()));
 }
 
 void BTSettingsWidget::load()
@@ -48,6 +49,7 @@ void BTSettingsWidget::save()
     BittorrentSettings::setTorrentDir(torrentEdit->url().url());
     BittorrentSettings::setTmpDir(tempEdit->url().url());
     BittorrentSettings::setPreAlloc(preallocBox->isChecked());
+    BittorrentSettings::setEnableUTP(utpBox->isChecked());
 
     BittorrentSettings::self()->writeConfig();
 }
@@ -60,6 +62,7 @@ void BTSettingsWidget::defaults()
     torrentEdit->setUrl(BittorrentSettings::torrentDir());
     tempEdit->setUrl(BittorrentSettings::tmpDir());
     preallocBox->setChecked(BittorrentSettings::preAlloc());
+    utpBox->setChecked(BittorrentSettings::enableUTP());
 }
 
 #include "btsettingswidget.moc"
--- a/kget/transfer-plugins/bittorrent/btsettingswidget.ui
+++ b/kget/transfer-plugins/bittorrent/btsettingswidget.ui
@@ -1,7 +1,8 @@
-<ui version="4.0" >
+<?xml version="1.0" encoding="UTF-8"?>
+<ui version="4.0">
  <class>BTSettingsWidget</class>
- <widget class="QWidget" name="BTSettingsWidget" >
-  <property name="geometry" >
+ <widget class="QWidget" name="BTSettingsWidget">
+  <property name="geometry">
    <rect>
     <x>0</x>
     <y>0</y>
@@ -9,140 +10,147 @@
     <height>363</height>
    </rect>
   </property>
-  <layout class="QVBoxLayout" name="verticalLayout" >
+  <layout class="QVBoxLayout" name="verticalLayout">
    <item>
-    <widget class="QGroupBox" name="networkGroupBox" >
-     <property name="sizePolicy" >
-      <sizepolicy vsizetype="Preferred" hsizetype="Preferred" >
+    <widget class="QGroupBox" name="networkGroupBox">
+     <property name="sizePolicy">
+      <sizepolicy hsizetype="Preferred" vsizetype="Preferred">
        <horstretch>0</horstretch>
        <verstretch>0</verstretch>
       </sizepolicy>
      </property>
-     <property name="title" >
+     <property name="title">
       <string>Network</string>
      </property>
-     <layout class="QFormLayout" name="formLayout" >
-      <property name="fieldGrowthPolicy" >
+     <layout class="QFormLayout" name="formLayout">
+      <property name="fieldGrowthPolicy">
        <enum>QFormLayout::AllNonFixedFieldsGrow</enum>
       </property>
-      <item row="0" column="0" >
-       <widget class="QLabel" name="portLabel" >
-        <property name="text" >
+      <item row="0" column="0">
+       <widget class="QLabel" name="portLabel">
+        <property name="text">
          <string>Port:</string>
         </property>
        </widget>
       </item>
-      <item row="0" column="1" >
-       <widget class="QSpinBox" name="portBox" >
-        <property name="maximum" >
+      <item row="0" column="1">
+       <widget class="QSpinBox" name="portBox">
+        <property name="maximum">
          <number>99000</number>
         </property>
-        <property name="value" >
+        <property name="value">
          <number>6881</number>
         </property>
        </widget>
       </item>
-      <item row="1" column="0" >
-       <widget class="QLabel" name="uploadLimitLabel" >
-        <property name="text" >
+      <item row="1" column="0">
+       <widget class="QLabel" name="uploadLimitLabel">
+        <property name="text">
          <string>Upload limit per transfer:</string>
         </property>
        </widget>
       </item>
-      <item row="1" column="1" >
-       <widget class="QSpinBox" name="uploadBox" >
-        <property name="specialValueText" >
+      <item row="1" column="1">
+       <widget class="QSpinBox" name="uploadBox">
+        <property name="specialValueText">
          <string>No Limit</string>
         </property>
-        <property name="suffix" >
+        <property name="suffix">
          <string>KiB</string>
         </property>
-        <property name="maximum" >
+        <property name="maximum">
          <number>10000</number>
         </property>
        </widget>
       </item>
-      <item row="2" column="0" >
-       <widget class="QLabel" name="downloadLabel" >
-        <property name="text" >
+      <item row="2" column="0">
+       <widget class="QLabel" name="downloadLabel">
+        <property name="text">
          <string>Download limit per transfer:</string>
         </property>
        </widget>
       </item>
-      <item row="2" column="1" >
-       <widget class="QSpinBox" name="downloadBox" >
-        <property name="specialValueText" >
+      <item row="2" column="1">
+       <widget class="QSpinBox" name="downloadBox">
+        <property name="specialValueText">
          <string>No Limit</string>
         </property>
-        <property name="suffix" >
+        <property name="suffix">
          <string>KiB</string>
         </property>
-        <property name="maximum" >
+        <property name="maximum">
          <number>100000</number>
         </property>
        </widget>
       </item>
+      <item row="3" column="1">
+       <widget class="QCheckBox" name="utpBox">
+        <property name="text">
+         <string>Enable UTP protocol</string>
+        </property>
+       </widget>
+      </item>
      </layout>
     </widget>
    </item>
    <item>
-    <widget class="QGroupBox" name="folderGroupBox" >
-     <property name="sizePolicy" >
-      <sizepolicy vsizetype="Preferred" hsizetype="Preferred" >
+    <widget class="QGroupBox" name="folderGroupBox">
+     <property name="sizePolicy">
+      <sizepolicy hsizetype="Preferred" vsizetype="Preferred">
        <horstretch>0</horstretch>
        <verstretch>0</verstretch>
       </sizepolicy>
      </property>
-     <property name="title" >
+     <property name="title">
       <string>Folders</string>
      </property>
-     <layout class="QFormLayout" name="formLayout_2" >
-      <item row="0" column="0" >
-       <widget class="QLabel" name="torrentFolderLabel" >
-        <property name="sizePolicy" >
-         <sizepolicy vsizetype="Minimum" hsizetype="Preferred" >
+     <layout class="QFormLayout" name="formLayout_2">
+      <item row="0" column="0">
+       <widget class="QLabel" name="torrentFolderLabel">
+        <property name="sizePolicy">
+         <sizepolicy hsizetype="Preferred" vsizetype="Minimum">
           <horstretch>0</horstretch>
           <verstretch>0</verstretch>
          </sizepolicy>
         </property>
-        <property name="text" >
+        <property name="text">
          <string>Default torrent folder:</string>
         </property>
        </widget>
       </item>
-      <item row="0" column="1" >
-       <widget class="KUrlRequester" name="torrentEdit" >
-        <property name="sizePolicy" >
-         <sizepolicy vsizetype="Fixed" hsizetype="Minimum" >
+      <item row="0" column="1">
+       <widget class="KUrlRequester" name="torrentEdit">
+        <property name="sizePolicy">
+         <sizepolicy hsizetype="Minimum" vsizetype="Fixed">
           <horstretch>0</horstretch>
           <verstretch>0</verstretch>
          </sizepolicy>
         </property>
-        <property name="filter" >
+        <property name="filter">
          <string/>
         </property>
        </widget>
       </item>
-      <item row="1" column="0" >
-       <widget class="QLabel" name="tempFolderLabel" >
-        <property name="text" >
+      <item row="1" column="0">
+       <widget class="QLabel" name="tempFolderLabel">
+        <property name="text">
          <string>Default temporary folder</string>
         </property>
        </widget>
       </item>
-      <item row="1" column="1" >
-       <widget class="KUrlRequester" name="tempEdit" >
-        <property name="sizePolicy" >
-         <sizepolicy vsizetype="Fixed" hsizetype="Minimum" >
+      <item row="1" column="1">
+       <widget class="KUrlRequester" name="tempEdit">
+        <property name="sizePolicy">
+         <sizepolicy hsizetype="Minimum" vsizetype="Fixed">
           <horstretch>0</horstretch>
           <verstretch>0</verstretch>
          </sizepolicy>
         </property>
        </widget>
       </item>
-      <item row="2" column="1" >
-       <widget class="QCheckBox" name="preallocBox" >
-        <property name="text" >
+      <item row="2" column="1">
+       <widget class="QCheckBox" name="preallocBox">
+        <property name="text">
          <string>Pre-allocate disk space</string>
         </property>
        </widget>
@@ -152,10 +160,10 @@
    </item>
    <item>
     <spacer>
-     <property name="orientation" >
+     <property name="orientation">
       <enum>Qt::Vertical</enum>
      </property>
-     <property name="sizeHint" stdset="0" >
+     <property name="sizeHint" stdset="0">
       <size>
        <width>20</width>
        <height>40</height>
--- a/kget/transfer-plugins/bittorrent/bttransfer.cpp
+++ b/kget/transfer-plugins/bittorrent/bttransfer.cpp
@@ -31,7 +31,8 @@
 #include <util/log.h>
 #include <peer/authenticationmonitor.h>
 #include <interfaces/trackerinterface.h>
-#include <btversion.h>
+#include <utp/utpserver.h>
+#include <version.h>
 
 #include <KDebug>
 #include <KLocale>
@@ -200,9 +201,7 @@ void BTTransfer::newDestResult()
 void BTTransfer::stop()
 {
     if (m_movingFile)
-    {
         return;
-    }
 
     if (m_ready)
     {
@@ -235,7 +234,7 @@ void BTTransfer::load(const QDomElement
 {
     Transfer::load(element);
 
-    if((m_totalSize == m_downloadedSize) && (m_totalSize != 0))
+    if ((m_totalSize == m_downloadedSize) && (m_totalSize != 0))
     {
         setStatus(Job::Stopped, i18nc("transfer state: finished", "Finished"), SmallIcon("dialog-ok"));
     }
@@ -254,7 +253,9 @@ void BTTransfer::load(const QDomElement
 
 void BTTransfer::setPort(int port)
 {
-    bt::Globals::instance().getServer().changePort(port);
+    bt::Globals::instance().getTCPServer().changePort(port);
+    if (BittorrentSettings::enableUTP())
+        bt::Globals::instance().getUTPServer().changePort(port + 1);
 }
 
 void BTTransfer::setSpeedLimits(int ulLimit, int dlLimit)
@@ -269,16 +270,14 @@ void BTTransfer::setSpeedLimits(int ulLi
 void BTTransfer::addTracker(const QString &url)
 {
     kDebug(5001);
-    if(torrent->getStats().priv_torrent)
-    {
-	KMessageBox::sorry(0, i18n("Cannot add a tracker to a private torrent."));
-	return;
+    if(torrent->getStats().priv_torrent) {
+        KMessageBox::sorry(0, i18n("Cannot add a tracker to a private torrent."));
+        return;
     }
 
-    if(!KUrl(url).isValid())
-    {
-	KMessageBox::error(0, i18n("Malformed URL."));
-	return;
+    if(!KUrl(url).isValid()) {
+       KMessageBox::error(0, i18n("Malformed URL."));
+       return;
     }
 
     torrent->getTrackersList()->addTracker(url,true);
@@ -449,16 +448,17 @@ void BTTransfer::btTransferInit(const KU
     bt::SetClientInfo("KGet", 2, KDE_VERSION_MINOR, KDE_VERSION_RELEASE, bt::NORMAL, "KG");//Set client info to KGet
 
     bt::Uint16 i = 0;
-    do
-    {
-        bt::Globals::instance().initServer(BittorrentSettings::port() + i);
+    while (!bt::Globals::instance().initTCPServer(BittorrentSettings::port() + i) && i < 10)
         i++;
-    }while (!bt::Globals::instance().getServer().isOK() && i < 10);
-
-    if (!bt::Globals::instance().getServer().isOK()) {
+    
+    if (i == 10) {
         setStatus(Job::Aborted, i18n("An error occurred...."), SmallIcon("document-preview"));
         return;
     }
+    if (BittorrentSettings::enableUTP()) {
+        while (!bt::Globals::instance().initUTPServer(BittorrentSettings::port() + i) && i < 10) //We don't care if it fails for now as UTP is experimental...
+            i++;
+    }
 
     QDir tmpDir(m_tmp + m_source.fileName().remove(".torrent"));
     if (tmpDir.exists())
@@ -496,6 +496,8 @@ void BTTransfer::btTransferInit(const KU
     catch (bt::Error &err)
     {
         m_ready = false;
+        torrent->deleteLater();
+        torrent = 0;
         setStatus(Job::Aborted, i18n("An error occurred...."), SmallIcon("document-preview"));
         KMessageBox::error(0, err.toString(), i18n("Error"));
     }
@@ -750,7 +752,7 @@ void BTTransfer::filesSelected()
 //     setTransferChange(Tc_TotalSize | Tc_DownloadedSize | Tc_Percent, true);
 }
 
-FileModel *BTTransfer::fileModel()//TODO korrektes file-model wenn nur eine datei im torrent!
+FileModel *BTTransfer::fileModel()//TODO correct file model for one-file-torrents
 {
     if (!m_fileModel)
     {
--- a/kget/transfer-plugins/bittorrent/bttransfer.h
+++ b/kget/transfer-plugins/bittorrent/bttransfer.h
@@ -1,6 +1,6 @@
 /* This file is part of the KDE project
 
-   Copyright (C) 2007 Lukas Appelhans <l.appelhans@gmx.de>
+   Copyright (C) 2007 - 2010 Lukas Appelhans <l.appelhans@gmx.de>
 
    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
--- a/kget/transfer-plugins/bittorrent/bttransferfactory.cpp
+++ b/kget/transfer-plugins/bittorrent/bttransferfactory.cpp
@@ -69,7 +69,7 @@ const QList<KAction *> BTTransferFactory
      BTTransferHandler * bttransfer = static_cast<BTTransferHandler *>(handler);
 
      QList<KAction*> actions;
-     if (bttransfer)
+     if (bttransfer && bttransfer->torrentControl())
      {
          KAction *openAdvancedDetailsAction = new KAction(KIcon("document-open"), i18n("&Advanced Details"), this);
  
--- a/kget/transfer-plugins/bittorrent/bttransferhandler.cpp
+++ b/kget/transfer-plugins/bittorrent/bttransferhandler.cpp
@@ -35,6 +35,8 @@ BTTransferHandler::~BTTransferHandler()
 
 void BTTransferHandler::createAdvancedDetails()
 {
+    if (!torrentControl())
+        return;
     kDebug(5001);
 
     if (!advancedDetails)
@@ -67,6 +69,8 @@ kt::Monitor* BTTransferHandler::torrentM
 
 void BTTransferHandler::createScanDlg()
 {
+    if (!torrentControl())
+        return;
     kDebug(5001);
     if (scanDlg)
     {
--- a/kget/transfer-plugins/bittorrent/kget_bittorrentfactory.kcfg
+++ b/kget/transfer-plugins/bittorrent/kget_bittorrentfactory.kcfg
@@ -14,6 +14,9 @@
     <entry name="Port" type="Int">
       <default>6881</default>
     </entry>
+    <entry name="EnableUTP" type="Bool">
+      <default>false</default>
+    </entry>
   </group>
   <group name="Dirs">
     <entry name="TorrentDir" type="String">
